src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -0,0 +1,409 @@
+
+  std::chrono::nanoseconds uptime(0u);
+  double uptime_seconds;
+  if (std::ifstream("/proc/uptime", std::ios::in) >> uptime_seconds)
+  {
+    uptime = std::chrono::nanoseconds(
+      static_cast<unsigned long long>(uptime_seconds*BILLION)
+    );
+  }
+
+  timeP -= uptime;
+
+  logger()->info("Creating Packetcapture instance");
+    setCapture(conf.getCapture());
+    setAnomimize(conf.getAnomimize());
+
+  if (conf.dumpIsSet()) {
+    setDump(conf.getDump());
+  }
+
+  setNetworkmode(conf.getNetworkmode());
+  addFilters(conf.getFilters());
+  addGlobalheader(conf.getGlobalheader());
+  CapStatus = (uint8_t) conf.getCapture();
+
+  auto t_filters_in = get_array_table<filters_table>("filters_map", 0, ProgramType::INGRESS);
+  auto t_filters_out = get_array_table<filters_table>("filters_map", 0, ProgramType::EGRESS);
+  filters_table ft_init;
+  ft_init.dst_port_flag = ft_init.src_port_flag = ft_init.network_filter_src_flag = ft_init.network_filter_dst_flag = ft_init.l4proto_flag = false;
+  t_filters_in.set(0x0, ft_init);
+  t_filters_out.set(0x0, ft_init);
+  writeHeader = true;
+  dt = "";
+}
+
+void Packetcapture::writeDump(const std::vector<uint8_t> &packet){
+
+  PacketJsonObject pj;
+  auto p = std::shared_ptr<Packet>(new Packet(*this, pj));
+  std::streamsize len;
+
+  if (dt == "") {
+    time_t now = time(0);
+    tm *ltm = localtime(&now);
+    std::stringstream tmstmp;
+    tmstmp << ltm->tm_mday << "-" << ltm->tm_mon << "-" << (ltm->tm_year + 1900) << "-" << ltm->tm_hour << ":" << ltm->tm_min << ":" << ltm->tm_sec;
+    dt = tmstmp.str();
+  }
+  
+  myFile.open("capture_"+ dt +".pcap", std::ios::binary | std::ios::app);
+
+  if (writeHeader == true) {
+    struct pcap_file_header *pcap_header = new struct pcap_file_header;
+    pcap_header->magic = global_header->getMagic();
+    pcap_header->version_major = global_header->getVersionMajor();
+    pcap_header->version_minor = global_header->getVersionMinor();
+    pcap_header->thiszone = global_header->getThiszone();   //timestamp are always in GMT
+    pcap_header->sigfigs = global_header->getSigfigs();
+    pcap_header->snaplen = filters->getSnaplen();
+    pcap_header->linktype = global_header->getLinktype();
+
+    myFile.write(reinterpret_cast<const char*>(pcap_header), sizeof(*pcap_header));
+    writeHeader = false;
+  }
+
+  /*
+   * Here the packet capture time offset must be added to the system boot time stored in epoch format.
+   * See line 61 of this file to see system boot time stored in epoch algorithm
+   */
+
+  struct timeval tp;
+  std::chrono::system_clock::time_point timeBase;
+  timeBase = timeP + temp_offset;
+  std::chrono::system_clock::duration tp_dur = timeBase.time_since_epoch();
+  
+  std::chrono::seconds s = std::chrono::duration_cast<std::chrono::seconds>(tp_dur);
+  tp_dur -= s;
+  std::chrono::microseconds us = std::chrono::duration_cast<std::chrono::microseconds>(tp_dur);
```

Comment:
Definitely better! Thanks @DavideAG.
Just last comment (yes, I know, I am maybe too fussy).
Wouldn't it be better to create a function that does the conversion? For example
```
template<typename Duration>
void to_timeval(Duration&& d, struct timeval & tv) {
    std::chrono::seconds const sec = std::chrono::duration_cast<std::chrono::seconds>(d);

    tv.tv_sec  = sec.count();
    tv.tv_usec = std::chrono::duration_cast<std::chrono::microseconds>(d - sec).count();
}
```
In this way you will just call `to_timeval(timeBase, &tp)` and it's done ;)

Reply:
Yes it's better using a function like this. Thank you for your help!